File: lib/features/azkar/data/repositories/azkar_repository_impl.dart

import 'package:islami_admin/features/azkar/data/datasources/azkar_remote_data_source.dart';
import 'package:islami_admin/features/azkar/domain/repositories/azkar_repository.dart';

class AzkarRepositoryImpl implements AzkarRepository {
  final AzkarRemoteDataSource remoteDataSource;

  AzkarRepositoryImpl({required this.remoteDataSource});

  @override
  Future<String> getAzkar() async {
    return await remoteDataSource.getAzkar();
  }

  @override
  Future<void> saveAzkar(String content) async {
    await remoteDataSource.saveAzkar(content);
  }
}

File: shared/src/commonMain/kotlin/com/islami/data/repository/AzkarRepositoryImpl.kt
package com.islami.data.repository

import com.islami.core.error.Result
import com.islami.data.remote.firebase.FirebaseFirestoreClient
import com.islami.domain.entities.Azkar
import com.islami.domain.repositories.AzkarRepository
import kotlinx.coroutines.flow.Flow

class AzkarRepositoryImpl(
    private val firestoreClient: FirebaseFirestoreClient
) : AzkarRepository {

    override suspend fun getAzkar(): Result<List<Azkar>> {
        return firestoreClient.getCollection(
            collection = "azkar",
            serializer = Azkar.serializer()
        )
    }

    override suspend fun getAzkarByCategory(category: String): Result<Azkar> {
        return firestoreClient.getDocument(
            collection = "azkar",
            documentId = category,
            serializer = Azkar.serializer()
        )
    }

    override suspend fun saveAzkar(azkar: Azkar): Result<Unit> {
        return firestoreClient.setDocument(
            collection = "azkar",
            documentId = azkar.id,
            data = azkar
        )
    }

    override fun observeAzkar(): Flow<Result<List<Azkar>>> {
        return firestoreClient.observeCollection(
            collection = "azkar",
            serializer = Azkar.serializer()
        )
    }
}
